File: javascript/levenshtein/levenshtein.js
function ditstance(source, target) {
    var src_length = source.length + 1;
    var tgt_length = target.length + 1;
    
    if (src_length == 1) {
        return tgt_length - 1;
    }
    if (tgt_length == 1) {
        return src_length - 1;
    }

    var matrix  = new Array();
    var row = new Array();
    for (i = 0; i < tgt_length; i++){
        row[i] = i;
    }
    matrix[0] = row;
    for (i = 1; i < src_length; i++) {
        var row = new Array();
        row[0] = i;
        for (j = 1; j < tgt_length; j++) {
            row[j] = 0;
        }
        matrix[i] = row;
    }

    for (i = 1; i < src_length; i++) {
        var src_char = source[i-1];
        for (j = 1; j < tgt_length; j++) {
            var tgt_char = target[j-1];
            var cost;
            if (src_char === tgt_char) {
                cost = 0;
            } else {
                cost = 1;
            }
            var above = matrix[i-1][j]+1;
            var left = matrix[i][j-1]+1;
            var diag = matrix[i-1][j-1]+cost;
            var value = Math.min(above, left, diag);
            matrix[i][j] = value;
        }
    }
    return matrix[src_length - 1][tgt_length - 1];
}

File: java/algorithm/src/com/xiachufang/Levenshtein.java
package com.xiachufang;

public class Levenshtein {
    public static int ditstance(String source, String target) {
        int src_length = source.length() + 1;
        int tgt_length = target.length() + 1;

        if (src_length == 1) {
            return tgt_length - 1;
        }

        if (tgt_length == 1) {
            return src_length - 1;
        }

        int[][] matrix = new int[src_length][tgt_length];
        for (int i = 0; i < src_length; i++) {
            matrix[i][0] = i;
        }
        for (int i = 0; i < tgt_length; i++) {
            matrix[0][i] = i;
        }

        for (int i = 1; i < src_length; i++) {
            char src_char = source.charAt(i-1);
            for (int j = 1; j < tgt_length; j++) {
                char tgt_char = target.charAt(j-1);
                int cost = (src_char == tgt_char)?0:1;
                int above = matrix[i-1][j]+1;
                int left = matrix[i][j-1]+1;
                int diag = matrix[i-1][j-1]+cost;
                int value = Math.min(above, Math.min(left, diag));
                matrix[i][j] = value;
            }
        }
        
        return matrix[src_length - 1][tgt_length - 1];
    }
}
